File: app/src/main/java/com/yalantis/api/LogOutWhenSessionExpired.java
package com.yalantis.api;

import android.content.Context;
import android.widget.Toast;

import com.yalantis.App;
import com.yalantis.R;
import com.yalantis.api.body.ErrorBody;

import retrofit2.adapter.rxjava.HttpException;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action0;
import rx.functions.Func1;

/**
 * Created by voltazor on 20/06/16.
 */
public class LogOutWhenSessionExpired implements Func1<Observable<? extends Throwable>, Observable<?>> {

    private Context mContext;

    public LogOutWhenSessionExpired(Context context) {
        this.mContext = context;
    }

    @Override
    public Observable<?> call(Observable<? extends Throwable> observable) {
        return observable.observeOn(AndroidSchedulers.mainThread()).flatMap(new Func1<Throwable, Observable<?>>() {

            @Override
            public Observable<?> call(final Throwable throwable) {
                if (throwable instanceof HttpException) {
                    ErrorBody errorBody = ErrorBody.parseError(((HttpException) throwable).response());
                    if (errorBody != null) {
                        final int code = errorBody.getCode();
                        if (code == ErrorBody.INVALID_TOKEN) {
                            return Observable.empty().observeOn(AndroidSchedulers.mainThread()).doOnCompleted(new Action0() {
                                @Override
                                public void call() {
                                    Toast.makeText(mContext, R.string.your_session_expired, Toast.LENGTH_SHORT).show();
                                    App.logOut(mContext);
                                }
                            });
                        }
                    }
                }
                // If cannot be handled - pass the original error through.
                return Observable.error(throwable);
            }

        });
    }

}

File: app/src/main/java/com/yalantis/api/LogOutWhenSessionExpired.kt
package com.yalantis.api

import android.content.Context
import android.widget.Toast
import com.yalantis.R
import com.yalantis.api.body.ErrorBody
import com.yalantis.manager.SharedPrefManager
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function
import retrofit2.HttpException

/**
 * Created by voltazor on 20/06/16.
 */
class LogOutWhenSessionExpired(private val mContext: Context) : Function<Observable<out Throwable>, Observable<Any>> {

    override fun apply(observable: Observable<out Throwable>): Observable<Any> {
        return observable.observeOn(AndroidSchedulers.mainThread()).flatMap(Function<Throwable, Observable<*>> { throwable ->
            if (throwable is HttpException) {
                val errorBody = ErrorBody.parseError(throwable.response())
                if (errorBody != null) {
                    val code = errorBody.code
                    if (code == ErrorBody.INVALID_TOKEN) {
                        return@Function Completable.complete().observeOn(AndroidSchedulers.mainThread()).doOnComplete {
                            Toast.makeText(mContext, R.string.your_session_expired, Toast.LENGTH_SHORT).show()
                            logOut()
                        }.toObservable<Any>()
                    }
                }
            }
            // If cannot be handled - pass the original error through.
            Observable.error<Any>(throwable)
        })
    }

    fun logOut() {
        SharedPrefManager.clear()
    }
}
